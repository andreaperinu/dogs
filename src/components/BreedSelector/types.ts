import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
