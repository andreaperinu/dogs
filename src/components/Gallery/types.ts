import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
