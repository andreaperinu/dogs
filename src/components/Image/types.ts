import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
