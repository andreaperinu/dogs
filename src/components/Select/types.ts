import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
