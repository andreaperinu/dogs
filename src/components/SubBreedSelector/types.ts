import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
