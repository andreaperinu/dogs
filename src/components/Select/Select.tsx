import classNames from 'classnames';
import * as A from 'fp-ts/lib/Array';
import * as B from 'fp-ts/lib/boolean';
import { constVoid, pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React, { memo, useCallback, useMemo } from 'react';

import { noopComponent } from '@/utils/utils';

import { SelectStyled } from './Select.styled';
import { ISelectItem, ISelectProps } from './types';

export const BaseSelect = ({
  maybeValue, maybeOptions, className, onChange, disabled = false, maybeText = O.none,
}: ISelectProps) => {
  const [showOptions, setShowOptions] = React.useState(false)

  const onItemSelect = useCallback((item: ISelectItem) => {
    setShowOptions(false)
    onChange(O.fromNullable(item))
  }, [onChange])

  const onShowOptionsHandler = useCallback(() => pipe(
    !disabled,
    B.fold(constVoid, () => setShowOptions((prev) => !prev))
  ), [disabled])

  const renderedValue = useMemo(() => pipe(
    maybeValue,
    O.map(({ value }) => value),
    O.getOrElse(() => ''),
    v => <div className={classNames('value', className)} onClick={onShowOptionsHandler}>{v}</div>
  ), [maybeValue, onShowOptionsHandler])

  const renderedOptions = useMemo(() => pipe(
    maybeOptions,
    O.map(A.map(option => (
      <li key={option.id} value={option.id} onClick={() => onItemSelect(option)}>
        {option.value}
      </li>
    ))),
    O.map(O.fromPredicate(A.isNonEmpty)),
    O.chain(O.map(items => <ul className="list">{items}</ul>)),
    O.getOrElse(noopComponent),
  ), [maybeOptions, onItemSelect])

  const renderedText = useMemo(() => pipe(
    maybeText,
    O.map(text => <div className="text" onClick={onShowOptionsHandler}>{text}</div>),
    O.getOrElse(noopComponent)
  ), [maybeText, onShowOptionsHandler])

  return (
    <SelectStyled showOptions={showOptions} disabled={disabled}>

      <div className="wrapper">
        {renderedValue}
        {renderedOptions}
      </div>

      {renderedText}
    </SelectStyled>
  )
}

export const Select = memo(BaseSelect)
