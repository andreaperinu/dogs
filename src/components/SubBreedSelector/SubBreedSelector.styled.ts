import styled, { css } from 'styled-components';

import { ISubBreedSelectorStyledProps } from './types';

export const SubBreedSelectorStyled = styled.div<ISubBreedSelectorStyledProps>`
${({ disabled }) => css`

		h2 {
			${disabled && css` opacity: 0.6; `}
		}
	`}
`
