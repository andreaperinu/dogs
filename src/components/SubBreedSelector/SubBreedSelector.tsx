import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { useCallback } from 'react';

import {
	maybeSubBreedBySelectItem, maybeSubBreedsToSelectItems, maybeSubBreedToSelectItem
} from '@/utils/utils';

import { Select } from '../Select/Select';
import { ISelectItem } from '../Select/types';
import { SubBreedSelectorStyled } from './SubBreedSelector.styled';
import { ISubBreedSelectorProps } from './types';

export const SubBreedSelector = ({
	maybeSelectedSubBreed, maybeBreed, disabled = false, onChange,
}: ISubBreedSelectorProps) => {
	const onChangeHandler = useCallback((maybeSelectItem: O.Option<ISelectItem>) => pipe(
		maybeSelectItem,
		maybeSubBreedBySelectItem,
		onChange,
	), [onChange])

	return (
		<SubBreedSelectorStyled disabled={disabled}>
			<h2>Then, select a sub-breed</h2>
			<Select
				disabled={disabled}
				maybeValue={pipe(maybeSelectedSubBreed, maybeSubBreedToSelectItem)}
				maybeOptions={pipe(maybeBreed, maybeSubBreedsToSelectItems)}
				onChange={onChangeHandler}
				data-test="component-select"
			/>
		</SubBreedSelectorStyled>
	)
}
