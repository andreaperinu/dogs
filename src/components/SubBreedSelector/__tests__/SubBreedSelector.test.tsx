import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React from 'react';

import { ISelectItem } from '@/components/Select/types';
import { IBreed } from '@/models/models';
import * as mockedBreeds from '@/tests/mockedBreeds.json';
import { findByDataTest, setup } from '@/tests/utils';
import { formatBreeds, maybeSubBreedToSelectItem } from '@/utils/utils';

import { SubBreedSelector as component } from '../SubBreedSelector';

const onChange = jest.fn()

const maybeBreed: O.Option<IBreed> = pipe(
	mockedBreeds,
	formatBreeds,
	A.head,
)

const maybeSelectedSubBreed: O.Option<string> = pipe(
	maybeBreed,
	O.map(({ subBreeds }) => subBreeds),
	O.chain(A.head),
)

const props: React.ComponentProps<typeof component> = {
	maybeBreed,
	maybeSelectedSubBreed,
	onChange,
}

describe('<BreedSelector />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props })
		expect(wrapper).toMatchSnapshot()
	})

	it('should set sub-breed', () => {
		const mockedMaybeSelectItem: O.Option<ISelectItem> = pipe(
			maybeSelectedSubBreed,
			maybeSubBreedToSelectItem,
		)

		const wrapper = setup({ component, props })
		const select = findByDataTest(wrapper, 'select')

		select?.simulate('change', mockedMaybeSelectItem)

		expect(onChange).toHaveBeenCalledWith(maybeSelectedSubBreed)
	})
})
