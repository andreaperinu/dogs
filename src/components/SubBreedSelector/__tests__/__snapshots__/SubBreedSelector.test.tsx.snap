// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreedSelector /> should match snapshot 1`] = `
<styled.div
  disabled={false}
>
  <h2>
    Then, select a sub-breed
  </h2>
  <Memo(BaseSelect)
    data-test="component-select"
    disabled={false}
    maybeOptions={
      Object {
        "_tag": "Some",
        "value": Array [],
      }
    }
    maybeValue={
      Object {
        "_tag": "None",
      }
    }
    onChange={[Function]}
  />
</styled.div>
`;
